src/controllers/OAuthController.ts

```diff
@@ -0,0 +1,726 @@
+import express, { NextFunction, Request, RequestHandler, Response } from "express";
+import JWT from "jsonwebtoken";
+import csrf from "csurf";
+import moment from "moment";
+import Controller from "../interfaces/Controller";
+import Service from "../models/Service";
+import User from "../models/User";
+import AuthenticationService from "../services/AuthenticationService";
+import PrivacyPolicyService from "../services/PrivacyPolicyService";
+import UserService from "../services/UserService";
+import { pick } from "lodash";
+import ServiceError from "../utils/ServiceError";
+import { stringToServiceToken } from "../token/Token";
+import AuthorizeMiddleware, { IASRequest } from "../utils/AuthorizeMiddleware";
+import ConsentService from "../services/ConsentService";
+import PrivacyPolicyConsent from "../enum/PrivacyPolicyConsent";
+import { OAuthError } from "../utils/OAuthError";
+
+const getIdToken = (user: User, scope: string[], service: Service) => {
+  const claimNames = getClaimNames(scope);
+  const claims = pick(getUserClaims(user, claimNames), getAllowedClaims(service.dataPermissions));
+
+  const iat = Date.now();
+
+  const token = {
+    iss: "http://users.tko-aly.localhost/",
+    aud: service.serviceIdentifier,
+    iat,
+    exp: Date.now() / 1000 + 60 * 60 * 3,
+    ...claims,
+  };
+
+  return JWT.sign(token, process.env.JWT_SECRET ?? "");
```

Comment:
Signing with "" if no token is set seems dangerous. We should instead error and refuse to sign anything.